### ISSUE TITLE
Asymmetric behavior for queried domains in function definitions

### ISSUE THREAD

**Issue by npadmana** [chapel-lang/chapel#15877]
(This captures a discussion on SO -- https://stackoverflow.com/questions/62397173/what-is-the-best-way-to-check-ensure-that-two-arrays-have-the-same-domain-and-di/)

Consider the following definitions --  
```chpl
const Dom = {1..5, 1..5};
const BDom = newBlockDom(Dom);
var x : [Dom] int;
var y : [BDom] int;

proc test(a : [?Dom1] int, b : [Dom1] int) {
}
```

As @bradcray pointed out, the call `test(x,y)` compiles correctly, while `test(y,x)` does not. This is because of how anonymous domains get treated.

The question is whether this is desirable, or `test(x,y)` should also not compile due to a mismatch in distributions.

**Comment by bradcray** [comment]
I tend to agree that the current behavior is more confusing than helpful for Nikhil's case.

A potential change to what we have today would be to only have array formals which are specified with an anonymous default rectangular domain be blind to the actual array argument's distribution and to treat any named or distributed domain specified within the formal type specifier as a constraint.  Thus:

```chapel
proc foo(X: [1..n] int) { ... }
```

would permit any n-element array of ints to be passed in regardless of distribution whereas:

```chapel
const D = {1..n};
proc foo(X: [D] int) { ... }
```

would require the actual to be compatible with D (specifically it would need to be non-distributed).  This is an approach we've discussed before and it would arguably address Nikhil's case by interpreting the use of the named queried domain for `b` as a hard constraint on it.

I'm not certain how hard this would be to implement.

Note that this would be a breaking language change.
